bootstrap/kubeadm/internal/cloudinit/kubeadm-bootstrap-script.sh

```diff
@@ -127,11 +127,24 @@ retry-command kubeadm join phase control-plane-prepare certs
 retry-command kubeadm join phase control-plane-prepare kubeconfig
 retry-command kubeadm join phase control-plane-prepare control-plane
 # {{ end }}
+
+# Check if the ControlPlaneKubeletLocalMode feature-gate is enabled
+kubectl get configmap -n kube-system kubeadm-config -o jsonpath='{.data.ClusterConfiguration}' | \
+  grep -q 'ControlPlaneKubeletLocalMode: true' > /dev/null && CONTROLPLANE_KUBELET_LOCAL_MODE="true" || CONTROLPLANE_KUBELET_LOCAL_MODE="false"
+
 retry-command kubeadm join phase kubelet-start
-# {{ if .ControlPlane }}
-try-or-die-command kubeadm join phase control-plane-join etcd
-retry-command kubeadm join phase control-plane-join update-status
-retry-command kubeadm join phase control-plane-join mark-control-plane
-# {{ end }}
+
+if [[ "${CONTROLPLANE_KUBELET_LOCAL_MODE}" == "true" ]]; then
+  # Run kubeadm join and skip all already executed phases
+  # kubeadm join relies on hidden phases when CONTROLPLANE_KUBELET_LOCAL_MODE. These cannot be run
+  # externally.
+  try-or-die-command kubeadm join \
+    --skip-phases preflight,control-plane-prepare/download-certs,control-plane-prepare/certs,control-plane-prepare/kubeconfig,control-plane-prepare/control-plane,kubelet-start
+else
+  # {{ if .ControlPlane }}
+  try-or-die-command kubeadm join phase control-plane-join etcd
+  retry-command kubeadm join phase control-plane-join mark-control-plane
+  # {{ end }}
+fi
```

Comment:
there is another phase at the end:
```
wait-control-plane     EXPERIMENTAL: Wait for the control plane to start
```
as i listed here:
https://github.com/kubernetes-sigs/cluster-api/pull/10983#discussion_r1700311886

https://github.com/kubernetes/kubernetes/blob/f8d5b2074c9b3a1629ef085b9e17a5370f393732/cmd/kubeadm/app/cmd/phases/join/waitcontrolplane.go#L63
but it comes from another alpha FG that can be skipped in the CAPI exp-join-retry IMO.

Reply:
but i guess it depends on the timeline of exp-retry-join vs features.WaitForAllControlPlaneComponents graudation.
i would assume that features.WaitForAllControlPlaneComponents will graduate faster than the removal of exp-retry-join, so might be worth adding it here.

it's a no-op if features.WaitForAllControlPlaneComponents is disabled.

Reply:
jep. I simplified it for now to just run `kubeadm join` for all leftover phases, so no retries on `mark-control-plane`, but with the Pro of running all hidden and potentially enabled and new phases.